pages/docs/couple-your-code/couple-your-code-gradient-data.md

```diff
@@ -0,0 +1,185 @@
+summary: "So far, our mesh contains only data. This is sufficient for most of the numerical methods that preCICE offers. For nearest-neighbor-gradient mapping, however, preCICE also requires additional gradient data information. In this step, you learn how to add gradient data to the mesh."
+---
+
+{% version 2.4.0 %}
+This feature is available since version 2.4.0.
+{% endversion %}
+
+When using `nearest-neighbor-gradient` mapping, we require coupling data and additional gradient data. We have seen in [Step 3](couple-your-code-mesh-and-data-access.html) how to write data to the mesh.
+Now, we will learn how to write gradient data to the mesh.
+
+For this purpose, we use the following API methods:
+
+```cpp
+bool isDataGradientRequired(int dataID);
+
+void writeScalarGradientData (
+    int             dataID,
+    int             valueIndex,
+    const double*   gradientValues  );
+
+void writeBlockScalarGradientData (
+      int           dataID,
+      int           size,
+      const int*    valueIndices,
+      const double* gradientValues  );
+      
+void writeVectorGradientData (
+    int             dataID,
+    int             valueIndex,
+    const double*   gradientValues,
+    bool            rowsFirst = false );
+
+void writeBlockVectorGradientData (
+      int           dataID,
+      int           size,
+      const int*    valueIndices,
+      const double* gradientValues,
+      bool          rowsFirst = false );
+```
+
+* `isDataGradientRequired` returns a boolean, indicates if the data corresponding to the ID `dataID` has gradient data.
+* `writeScalarGradientData` writes scalar-valued gradient data, derived in each spatial dimension to the coupling data structure.
+* `ẁriteBlockScalarGradintData` writes multiple scalar gradient data at once, for performance reasons.
+* `writeVectorGradientData` writes vector-valued gradient data to the coupling data structure. The matrix is entered as a 1D-array of each component differentiated first. The parameter `rowsFirst` allows to write the values of the matrix, such as the components are differentiated in the spatial dimensions first.
+* `ẁriteBlockVectorGradintData` writes multiple vector gradient data at once, for performance reasons.
+
+Let's consider an example for writing block vector gradient data corresponding to the vector data `v0 = (v0x, v0y) , v1 = (v1x, v1y), ... , vn = (vnx, vny)` differentiated in spatial directions x and y.
+Per default, the values are passed as following:
+
+```cpp
+( v0x_dx, v0x_dy, v0y_dx, v0y_dy,
+  v1x_dx, v1x_dy, v1y_dx, v1y_dy,
+  ... ,
+  vnx_dx, vnx_dy, vny_dx, vny_dy  )  
+```
+
+When using the `rowsFirst` parameter, the following format is required:
+
+```cpp
+( v0x_dx, v0y_dx, v1x_dx, v1y_dx, ... , vnx_dx, vny_dx,
+  v0x_dy, v0y_dy, v1x_dy, v1y_dy, ... , vnx_dy, vny_dy  )
+```
+
+{% note %}
+Gradient data must be added to an existing data ID. Therefore, we must create data first, then the corresponding gradient data.
+{% endnote %}
```

Comment:
```suggestion
```

The user does not create a data field. preCICE does when specified in the config. You probably mean that the user needs to have a valid data ID. But this should be clear anyway already (from step 3). Nothing to specify here in my opinion.